File: python/word_break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = { len(s): True }
        def backtrack(index: int) -> bool:
            if index not in memo:
                for word in wordDict:
                    if s.find(word, index) == index:
                        if backtrack(index + len(word)):
                            memo[index] = True
                            break
                if index not in memo:
                    memo[index] = False
            return memo[index]
        return backtrack(0)

File: java/wordBreak.java
class Solution {
    int curIndex = 0;
    String s = "";
    List<String> wordDict = new ArrayList();
    Map<Integer, Boolean> memo = new HashMap();
    
    public boolean wordBreak(String s, List<String> wordDict) {
        this.s = s;
        this.wordDict = wordDict;
        memo.put(s.length(), true);
        return backtrack();
    }
    
    private boolean backtrack() {
        if (!memo.keySet().contains(curIndex)) {
            for (String word : wordDict) {
                if (word.length() <= s.length() - curIndex) {
                    if (s.substring(curIndex, curIndex + word.length()).equals(word)) {
                        curIndex += word.length();
                        if (backtrack()) {
                            return true;
                        }
                        curIndex -= word.length();
                    }
                    
                }
            }
            memo.put(curIndex, false);
        }
        return memo.get(curIndex);
    }
}